### DISCUSSION TITLE
Why did we use VRFMock contract even for sepolia chain?

### DISCUSSION THREAD

**Discussion by msyuvaraj** [Cyfrin/foundry-full-course-cu#1077] [Q&A] [answered]
In Lesson 9 ,
in interactions script in the addConsumer Function why did we broadcast sepolia vrf address in the  vrfMock contract.Aren't we supposed to use different vrfContract for different chain?

`contract AddConsumer is Script {
    function addConsumer(address vrfCoordinator , address contractToAddToVrf , uint64 subId) public {
        vm.startBroadcast();
        VRFCoordinatorV2Mock(vrfCoordinator).addConsumer(
            subId,
            contractToAddToVrf
        );
        vm.stopBroadcast();
    }

    function addConsumerUsingConfig(address mostRecentlyDeployed) public {
        HelperConfig helperConfig = new HelperConfig();
        (, , address vrfCoordinator, ,uint64 subId ,, ) = helperConfig.activeNetworkConfig();
        addConsumer(vrfCoordinator, mostRecentlyDeployed, subId);

    }
    
    function run() external {
         address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "Raffle",
            block.chainid
        );
        addConsumerUsingConfig(mostRecentlyDeployed);
    }
}`

**Comment by Niferu** [comment]
Hello @msyuvaraj 

You are right, we should be using different VRFV2 contracts, mock one for local network and already existing (deployed) one for Sepolia for example. And in this code we are doing it using correct vrfV2 contract Address.

If you have doubts regarding this line:
```solidity
        vm.startBroadcast(deployerKey);
        VRFCoordinatorV2Mock(vrfCoordinator).addConsumer(subId, contractToAddToVrf);
        vm.stopBroadcast();
```

We are indeed using `VRFCoordinatorV2Mock` but it is only for interface purposes as you can see we are wrapping correct vrfV2 contract address with it, so it will have correct interface functions. You can wrap it with VRFCoordinatorV2 interface too and it should work the same. The important part is in our helper contract which is picking correct vrf address for us, which depends on network.

**Reply by sunnyStefi** [reply]
The function that's called by the contract it's the one that's referred by the type of the contract `VRFCoordinatorV2Mock` or `VRFCoordinatorV2`: in this case they both have the same signature and the EVM does not check the type of the address for us

**Reply by Niferu** [reply]
@sunnyStefi  You exactly, that is just an interface matter and correct contract address.

**Reply by sunnyStefi** [reply]
Hei @Nlferu, thanks for replying! 
It's true that inside the interaction we are passing the correct address from the `HelperConfig.s.sol`, but that's not enough to perform the correct function `addConsumer`.

Here the function that will be executed in this case will be the one of the **mock**:

```solidity 
VRFCoordinatorV2Mock(vrfCoordinator).addConsumer(subscriptionId, lottery);
```

where here the function that will be executed will be the one of the **VRFCoordinatorV2** :

```solidity 
VRFCoordinatorV2(vrfCoordinator).addConsumer(subscriptionId, lottery)
```

Even tough the address of the vrfCoordinator is the one we intend to be. You can also call the mock `addConsumer` function inside the address of the real vrfCoordinator one and viceversa (as we can do in @msyuvaraj case) since the functionality they perform is the same.

**Reply by msyuvaraj** [reply]
Tq guys @Nlferu @sunnyStefi ,understood the technicalities behind it.

**Reply by Niferu** [reply]
Hi @sunnyStefi 

Yes that is exactly what I have said :) we are just parsing correct interface to proper contract address.

**Comment by sunnyStefi** [comment, marked as answer]
Hi @msyuvaraj! 

Both `VRFCoordinatorV2` and `VRFCoordinatorV2Mock` will work for sepolia network since they have a function with the same signature that behaves in a very similar way. They will all add a consumer to the VRF's state variable `s_consumers` for the given **subscriptionId**, and both emit events that are recognised by the Chainlink Node: `ConsumerAdded(_subId, _consumer)` for the mock and `SubscriptionConsumerAdded(subId, consumer)`.

We have a similar example inside `CreateSubscription` interaction: if you search inside both contracts you'll see their exact behaviour except that the `VRFCoordinatorV2` will add some extra configuration. 

Stefi
